### ISSUE TITLE
Drive-by feedback

### ISSUE THREAD

**Issue by mbaldessari** [Kubeinit/kubeinit#171]
Hi all,

just a super-quick drive-by feedback as I spent some time today trying kubeinit (on centos 8 hypervisor).
The first error I got was:
TASK [../../roles/kubeinit_okd : update packages] ******************************************************************
fatal: [hypervisor-01]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: kex_exchange_identification: Connection closed by remote host", "unreachable": true}

The reason for that is that the code has many sections like the following:
~~~
- name: Deploy a CentOS based guest
  block:
    - name: wait for {{ kubeinit_deployment_node_name }} to boot
      ansible.builtin.wait_for:
        port: 22
        host: "{{ hostvars[kubeinit_deployment_node_name].ansible_host }}"
        search_regex: OpenSSH
        delay: 10
~~~

These only check that openssh is listening to port 22, they do not guarantee that ansible is able to login. I had to add the following to make it past this issue:
~~~
--- a/kubeinit/roles/kubeinit_libvirt/tasks/20_check_nodes_up.yml
+++ b/kubeinit/roles/kubeinit_libvirt/tasks/20_check_nodes_up.yml
@@ -31,6 +31,13 @@
         # We can not resolve  by name from the hypervisors
         # ssh-keyscan -H {{ kubeinit_deployment_node_name }} >> ~/.ssh/known_hosts
         # ssh-keyscan -H {{ kubeinit_deployment_node_name }}.{{ kubeinit_inventory_cluster_domain }} >> ~/.ssh/known_hosts
+
+    - name: wait for {{ kubeinit_deployment_node_name }} to be accessible via ssh
+      ansible.builtin.wait_for_connection:
+        delay: 10
+        timeout: 600
+      delegate_to: "{{ hostvars[kubeinit_deployment_node_name].ansible_host }}"
+
   delegate_to: "{{ groups['hypervisor_nodes'][0] }}"
   tags:
     - provision_libvirt
~~~

Then I had a similar error but in a different spot (kubeinit/roles/kubeinit_okd/tasks/10_configure_service_nodes.yml).
Transforming the wait_for into wait_for_connection made it work for me.

On the next run it failed with:
~~~
TASK [../../roles/kubeinit_okd : verify that master nodes are ok] **********************************************************************************************************************************************************************
failed: [hypervisor-01] (item=okd-service-01) => {"ansible_loop_var": "cluster_node", "cluster_node": "okd-service-01", "msg": "Failed to connect to the host via ssh: kex_exchange_identification: Connection closed by remote host", "unreachable": true}
fatal: [hypervisor-01]: UNREACHABLE! => {"changed": false, "msg": "All items completed", "results": [{"ansible_loop_var": "cluster_node", "cluster_node": "okd-service-01", "msg": "Failed to connect to the host via ssh: kex_exchange_identification: Connection closed by remote host", "unreachable": true}]}
~~~

I was going to try and debug this as well but then docker.io started ratelimiting me and I ran out of time :)

TLDR: the wait_for tasks that check for openssh should be moved to wait_for_connection to avoid races

**Comment by ccamacho** [comment]
Hi @mbaldessari :) thanks a lot for the feedback, you are right I missed fixing this for a long time. It is already fixed in #172 also in the case of the docker registry PITA, you can export two variables like:

export KUBEINIT_COMMON_DOCKER_USERNAME=mydockeruser
export KUBEINIT_COMMON_DOCKER_PASSWORD=mydockerpass

This is defined/used [here](https://github.com/Kubeinit/kubeinit/blob/master/kubeinit/hosts/common.yml) 
And then execute your deployment command as usual.

**Comment by mbaldessari** [comment]
Thanks for the fix! I'll close this one now
